src/fluree/db/session.cljc

```diff
@@ -79,71 +65,150 @@
-              db-schema       (schema/schema-map db)        ;; returns async chan
-              db-settings     (schema/setting-map db)       ;; returns async chan
-              db*             (assoc db :schema (<? db-schema)
-                                        :settings (<? db-settings))]
-          (async/put! pc db*))
-        (catch* e
-                (async/put! pc e))))
-    pc))
+(defn ready?
+  [db-info]
+  (= :ready
+     (-> db-info :status keyword)))
+
+(defn load-ledger-info
+  [conn network dbid]
+  (go-try
+   (let [ledger-info (<? (ops/ledger-info-async conn [network dbid]))]
+     (if (empty? ledger-info)
+       (throw (ex-info (str "Ledger " network "/" dbid
+                            " is not found on this ledger group.")
+                       {:status 400
+                        :error  :db/unavailable}))
+       (if-not (ready? ledger-info)
+         (throw (ex-info (str "Ledger " network "/" dbid
+                              " is not currently available. Status is: "
+                              (:status ledger-info) ".")
+                         {:status 400
+                          :error  :db/unavailable}))
+         ledger-info)))))
+
+(defn load-current-db
+  [conn {:keys [network dbid] :as blank-db}]
+  (go-try
+   (let [{:keys [index], latest-block :block, :as ledger-info}
+         (<? (load-ledger-info conn network dbid))]
+     (when-let [indexed-db (<? (storage/reify-db conn network dbid blank-db index))]
+       (loop [db         indexed-db
+              next-block (-> indexed-db :block inc)]
+         (if (> next-block latest-block)
+           (let [schema   (<? (schema/schema-map db))
+                 settings (<? (schema/setting-map db))]
+             (swap! (:schema-cache db) empty)
+             (assoc db :schema schema, :settings settings))
+           (if-let [{:keys [flakes block t]}
+                    (<? (storage/read-block conn network dbid next-block))]
+             (recur (<? (dbproto/-with db block flakes))
+                    (inc next-block))
+             (throw (ex-info (str "Error reading block " next-block " for ledger: "
+                                  network "/" dbid ".")
+                             {:status 500, :error :db/unexpected-error})))))))))
+
+(defn reload-and-respond
+  [conn {:keys [network dbid] :as blank-db} resp-ch]
+  (go
+    (try*
+     (let [latest-db (<? (load-current-db conn blank-db))]
+       (async/put! resp-ch latest-db)
+       latest-db)
+     (catch* e
+             (log/error e
+                        "Error loading latest database for ledger" network dbid
+                        "in session")
+             (async/put! resp-ch e)
+             nil))))
+
+(defn new-db-cache
+  [conn network dbid cur-db]
+  (let [cur-ch (chan)
+        cas-ch (chan)
+        clr-ch (chan)
+        rel-ch (chan)]
+    (go-loop [db cur-db]
+      (async/alt!
+        :priority true
```

Comment:
Loading the current-db is done for most queries. I've not messed around with alt! + priority much but just want to flag this as a potential performance issue running these various channels. It may be negligible, but would be nice to do a criterium/quick-bench on fdb/current-db call to make sure we haven't introduced anything significant.

Reply:
There was a performance issue that was only revealed in the merge branch. Because the requests to clear the db and/or get the current db was serialized, lots of requests to load the current db got stale dbs because they requested the current db, then a new block came in making that db stale, but the "clear/reload" db request wouldn't process until the old db loaded for everything that asked for it before the new block came in. This is what I was trying to fix with the priority `alt!`.  This was fine here (except for two failed tests that I fixed with `syncTo`), but lots of tests failed on the 2.0 branch merging in these changes because we are running a lot of tests in parallel there. 

The code here was technically correct in that the stale db was current at the time it was requested, but so many tests failed in the merge branch because of race conditions. We could have used `syncTo` everywhere there to fix the tests, but I felt like there could be similar workflows that would also fail in production, and telling people to update all their queries with `syncTo` seemed like the wrong move.

After trying a bunch of things, I settled on going back to the atom based db cache and caching the promise channels instead of the db itself, but now the code waits on the db to load in a background go block and puts it on the cached channel in the case of success, and immediately clears the cache if there was a load error, so we don't cache errors and can recover from transient ones.

Reply:
The parallelism was also why everything would pass when I ran the tests one at a time, but there were tons of cascading failures when I ran the whole test suite (on the 2.0 merge branch). All the tests pass with the most recent commit.